File: python/3403.find-the-lexicographically-largest-string-from-the-box-i.py
class Solution:
    def answerString(self, word: str, numFriends: int) -> str:
        if numFriends == 1:
            return word

        n = len(word)
        win = n - numFriends + 1
        i = 0
        j = 1
        k = 0

        while j + k < n:

            if word[i + k] == word[j + k]:
                k += 1
            elif word[i + k] > word[j + k]:
                j = j + k + 1
                k = 0
            else:
                i = max(i + k + 1, j)
                j = i + 1
                k = 0

        return word[i: i + win]

File: golang/3403.find-the-lexicographically-largest-string-from-the-box-i.go
func answerString(word string, numFriends int) string {
	if numFriends == 1 {
		return word
	}

	n := len(word)
	win := n - numFriends + 1
	i, j, k := 0, 1, 0

	for j+k < n {
		if word[i+k] == word[j+k] {
			k++
		} else if word[i+k] > word[j+k] {
			j = j + k + 1
			k = 0
		} else {
			i = i + k + 1
			j = i + 1
			k = 0
		}
	}

	return word[i:min(i+win, n)]
}
